portfolio/src/main/webapp/script.js

```diff
@@ -42,3 +42,515 @@ function visitLink(s){
+                        return 3;
+                    }
+                }
+                return false;
+            }
+            else if(square.row == this.selected_square.row && 
+                square.col == this.selected_col + 2){
+                if(selected_piece.isWhite){
+                    if(!this.row[7][1].piece && !this.row[7][2].piece && !this.row[7][3].piece){
+                        selected_piece.moved = true;
+                        return 2;
+                    }
+                }
+                else{
+                    if(!this.row[0][1].piece && !this.row[0][2].piece && !this.row[0][3].piece){
+                        selected_piece.moved = true;
+                        return 3;
+                    }
+                }
+                return false;
+            }
+            return false;
+        }
+        else{
+            return false;
+        }
+    }
+
+    squareClicked(square){
+        // alert(`${square.row} ${square.col}`);
+
+        //if there is a previously selected square
+        if(this.selected_square){
+
+            //determine if move is legal
+            let legal = this.checkLegal(square, this.selected_square.piece);
+
+            //if the move is legal
+            if(legal == true){
+                //if there is a piece on the square
+                if(square.piece){
+                    //if the piece is the same color as the piece on the previously selected square
+                    if(square.piece.isWhite == this.selected_square.piece.isWhite){
+                        this.selected_square = square;
+                        console.log("Selected piece on: (" + square.row + ", " + square.col + ")");
+                        return;
+                    }
+                    //if the piece is not the same color (capture the piece)
+                    else{
+                        for(var ii = 0; ii < 32; ii++){
+                            if(this.pieceSquares[ii].row == square.row && this.pieceSquares[ii].col == square.col){
+                                this.pieceElems[ii].style.visibility = "hidden";
+                                // this.pieceSquares[ii] = null;
+                            }
+                        }
+                    }
+                }
+                //move the piece to the square
+                square.piece = this.selected_square.piece;
+                this.selected_square.piece = null;
+                for(var ii = 0; ii < this.pieceSquares.length; ii++){
+                    if(this.selected_square == this.pieceSquares[ii]){
+                        this.pieceSquares[ii] = square;
+                        this.pieceElems[ii].style.marginLeft = "-" + 
+                            ((8 - square.col)*75) + "px";
+                        this.pieceElems[ii].style.marginTop = "-" + 
+                            ((7 - square.row)*75) + "px";
+                    }
+                }
+
+                this.selected_square = null;
```

Comment:
This chunk has to happen for both regular moves and castling (and en passant).  For the non-regular moves it is accompanied by extra stuff but this base logic remains the same. I would either refactor it into a helper method. so it can be reused, or restructure the if statement surrounding this to make this only written once.